NeoML/include/NeoML/Dnn/Layers/IndRnnLayer.h

```diff
@@ -0,0 +1,128 @@
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+--------------------------------------------------------------------------------------------------------------*/
+
+#pragma once
+
+#include <NeoML/NeoMLDefs.h>
+#include <NeoML/Dnn/Dnn.h>
+#include <NeoML/Dnn/Layers/CompositeLayer.h>
+#include <NeoML/Dnn/Layers/DropoutLayer.h>
+#include <NeoML/Dnn/Layers/FullyConnectedLayer.h>
+
+namespace NeoML {
+
+// Forward declaration(s)
+class CIndRnnRecurrentLayer;
+
+// Independently Recurrent Neural Network (IndRNN): https://arxiv.org/pdf/1803.04831.pdf
+//
+// It's a simple recurrent unit with the following formula:
+//    Y_t = sigmoid( W * X_t + B + U * Y_t-1 )
+// Where:
+//    W and B are weights and free terms of the fully-connected layer (W * X_t is a matrix multiplication)
+//    U is a vector (U * Y_t-1 is an eltwise multiplication of 2 vectors of the same length)
+
+class NEOML_API CIndRnnLayer : public CCompositeLayer {
+	NEOML_DNN_LAYER( CIndRnnLayer )
+public:
+	explicit CIndRnnLayer( IMathEngine& mathEngine );
+
+	void Serialize( CArchive& archive ) override;
+
+	// Layer settings
+
+	// Sets the number of element in hidden layer and vector U
+	// 1 by default
+	int GetHiddenSize() const { return fc->GetNumberOfElements(); }
+	void SetHiddenSize( int hiddenSize );
+
+	// Sets dropout rate
+	// Dropout is applied to X_t and Y_t-1 before any other operations
+	// 0 by default (no dropout)
+	float GetDropoutRate() const { return inputDropout == nullptr ? 0 : inputDropout->GetDropoutRate(); }
+	void SetDropoutRate( float rate );
+
+	// Sets whether sequences must be processed in direct or reversed order
+	bool IsReverseSequence() const;
+	void SetReverseSequence( bool reverse );
+
+	// Trainable parameters
+	
+	// Input weights (matrix W from formula)
+	CPtr<CDnnBlob> GetInputWeights() const { return fc->GetWeightsData(); }
+	void SetInputWeights( const CDnnBlob* inputWeights ) { fc->SetWeightsData( inputWeights ); }
+
+	// Recurrent weights (vector U from formula)
+	CPtr<CDnnBlob> GetRecurrentWeights() const;
+	void SetRecurrentWeights( const CDnnBlob* recurrentWeights );
+
+	// Bias (vector B from formula)
+	CPtr<CDnnBlob> GetBias() const { return fc->GetFreeTermData(); }
+	void SetBias( const CDnnBlob* bias ) { fc->SetFreeTermData( bias ); }
+
+private:
+	CPtr<CDropoutLayer> inputDropout; // input dropout (if needed)
+	CPtr<CFullyConnectedLayer> fc; // Fully connected layer (W and B)
+	CPtr<CIndRnnRecurrentLayer> recurrent; // Recurrent part
+
+	void buildLayer();
+};
+
+// Recurrent part of IndRNN:
+//    Y_t = sigmoid( FC( X_t ) + U * Y_t-1 )
+//
+// For optimization purpose this class doesn't derive CRecurrentLayer
```

Comment:
purpose**s**; inherit instead of derive? Or **is derived from**, passive.

Reply:
Fixed